server/cmd/api/api/display.go

```diff
@@ -0,0 +1,616 @@
+	}
+
+	if restartChrome {
+		s.restartChromium(ctx)
+	}
+
+	return nil
+}
+
+// setResolutionXorgViaXrandr changes resolution for Xorg using xrandr (fallback when Neko is disabled)
+func (s *ApiService) setResolutionXorgViaXrandr(ctx context.Context, width, height, refreshRate int, restartChrome bool) error {
+	log := logger.FromContext(ctx)
+	display := s.resolveDisplayFromEnv()
+
+	// Build xrandr command - if refresh rate is specified, use the specific modeline
+	var xrandrCmd string
+	if refreshRate > 0 {
+		modeName := fmt.Sprintf("%dx%d_%d.00", width, height, refreshRate)
+		xrandrCmd = fmt.Sprintf("xrandr --output default --mode %s", modeName)
+		log.Info("using specific modeline", "mode", modeName)
+	} else {
+		xrandrCmd = fmt.Sprintf("xrandr -s %dx%d", width, height)
+	}
+
+	args := []string{"-lc", xrandrCmd}
+	env := map[string]string{"DISPLAY": display}
+	execReq := oapi.ProcessExecRequest{Command: "bash", Args: &args, Env: &env}
+	resp, err := s.ProcessExec(ctx, oapi.ProcessExecRequestObject{Body: &execReq})
+	if err != nil {
+		return fmt.Errorf("failed to execute xrandr: %w", err)
+	}
+
+	switch r := resp.(type) {
+	case oapi.ProcessExec200JSONResponse:
+		if r.ExitCode != nil && *r.ExitCode != 0 {
+			var stderr string
+			if r.StderrB64 != nil {
+				if b, decErr := base64.StdEncoding.DecodeString(*r.StderrB64); decErr == nil {
+					stderr = strings.TrimSpace(string(b))
+				}
+			}
+			if stderr == "" {
+				stderr = "xrandr returned non-zero exit code"
+			}
+			return fmt.Errorf("xrandr failed: %s", stderr)
+		}
+		log.Info("resolution updated via xrandr", "display", display, "width", width, "height", height)
+
+		if restartChrome {
+			s.restartChromium(ctx)
+		}
+		return nil
+	case oapi.ProcessExec400JSONResponse:
+		return fmt.Errorf("bad request: %s", r.Message)
+	case oapi.ProcessExec500JSONResponse:
+		return fmt.Errorf("internal error: %s", r.Message)
+	default:
+		return fmt.Errorf("unexpected response from process exec")
+	}
+}
+
+// setResolutionXvfb changes resolution for Xvfb by updating config and restarting services
+func (s *ApiService) setResolutionXvfb(ctx context.Context, width, height int, restartChrome bool) error {
+	log := logger.FromContext(ctx)
+	log.Info("updating Xvfb resolution requires restart", "width", width, "height", height)
+
+	// Update supervisor config to include environment variables
+	log.Info("updating xvfb supervisor config with new dimensions")
+	removeEnvCmd := []string{"-lc", `sed -i '/^environment=/d' /etc/supervisor/conf.d/services/xvfb.conf`}
+	removeEnvReq := oapi.ProcessExecRequest{Command: "bash", Args: &removeEnvCmd}
+	s.ProcessExec(ctx, oapi.ProcessExecRequestObject{Body: &removeEnvReq})
+
+	// Add the environment line with WIDTH and HEIGHT
+	addEnvCmd := []string{"-lc", fmt.Sprintf(`sed -i '/\[program:xvfb\]/a environment=WIDTH="%d",HEIGHT="%d",DPI="96",DISPLAY=":1"' /etc/supervisor/conf.d/services/xvfb.conf`, width, height)}
```

Comment:
this logic is sufficiently intricate where an e2e test might be beneficial. See https://github.com/onkernel/kernel-images/blob/main/server/e2e/e2e_chromium_test.go#L102-L219